gapis/resolve/dependencygraph2/graph_visualization/graph_structure.go

```diff
@@ -257,60 +265,111 @@ func (g *graph) makeStronglyConnectedComponentsByCommandTypeId() {
-		output += lines
+func (g *graph) dfs(currentNode *node, visited []bool, visitedNodes *[]*node) {
+	*visitedNodes = append(*visitedNodes, currentNode)
+	visited[currentNode.id] = true
+	neighbours := g.getSortedNeighbours(currentNode.outNeighbourIdToEdgeId)
+	for _, neighbour := range neighbours {
+		if !visited[neighbour.id] {
+			g.dfs(neighbour, visited, visitedNodes)
+		}
+	}
+
+	for _, subCommandNode := range currentNode.subCommandNodes {
+		if !visited[subCommandNode.id] {
+			g.dfs(subCommandNode, visited, visitedNodes)
+		}
 	}
-	return output
 }
 
-func (g *graph) getNodesInDotFormat() string {
-	output := ""
-	for _, currentNode := range g.nodeIdToNode {
-		lines := strconv.Itoa(currentNode.id) + "[label=" + currentNode.label + "]" + ";\n"
-		output += lines
+func (g *graph) joinNodesByFrame() {
+	visited := make([]bool, g.maxNodeId+1)
+	frameNumber := 1
+	nodes := g.getSortedNodes()
+	for _, currentNode := range nodes {
+		if !visited[currentNode.id] && currentNode.isEndOfFrame {
+			visitedNodes := []*node{}
+			g.dfs(currentNode, visited, &visitedNodes)
+			for _, visitedNode := range visitedNodes {
+				visitedNode.label = fmt.Sprintf("%s%d/%s", FRAME, frameNumber, visitedNode.label)
+			}
+			frameNumber++
+		}
 	}
-	return output
 }
 
-func (g *graph) getGraphInDotFormat() string {
-	output := "digraph g {\n"
-	output += g.getNodesInDotFormat()
-	output += g.getEdgesInDotFormat()
-	output += "}\n"
-	return output
+func (g *graph) joinNodesWithZeroDegree() {
+	for _, currentNode := range g.nodeIdToNode {
+		if (len(currentNode.inNeighbourIdToEdgeId) + len(currentNode.outNeighbourIdToEdgeId)) == 0 {
+			currentNode.label = UNUSED + "/" + currentNode.label
+		}
+	}
 }
 
-func (g *graph) getGraphInPbtxtFormat() string {
+func (g *graph) getSortedNodes() []*node {
 	nodes := []*node{}
 	for _, currentNode := range g.nodeIdToNode {
 		nodes = append(nodes, currentNode)
 	}
 	sort.Sort(nodeSorter(nodes))
+	return nodes
+}
 
-	output := ""
-	for _, currentNode := range nodes {
-		lines := "node {\n"
-		lines += "name: \"" + currentNode.label + "\"\n"
-		lines += "op: \"" + currentNode.label + "\"\n"
+func (g *graph) getSortedNeighbours(neighbourIdToEdgeId map[int]int) []*node {
+	neighbours := []*node{}
+	for neighbourId := range neighbourIdToEdgeId {
+		neighbours = append(neighbours, g.nodeIdToNode[neighbourId])
+	}
+	sort.Sort(nodeSorter(neighbours))
+	return neighbours
+}
 
-		neighbours := []*node{}
-		for neighbourId := range currentNode.inNeighbourIdToEdgeId {
-			neighbours = append(neighbours, g.nodeIdToNode[neighbourId])
+func (g *graph) getEdgesInDotFormat() []byte {
+	nodes := g.getSortedNodes()
+	var output bytes.Buffer
+	for _, currentNode := range nodes {
+		inNeighbours := g.getSortedNeighbours(currentNode.inNeighbourIdToEdgeId)
+		for _, neighbour := range inNeighbours {
+			output.WriteString(fmt.Sprintf("%d -> %d;\n", neighbour.id, currentNode.id))
```

Comment:
'fmt.Fprintf` will let you write directly to the `bytes.Buffer`, rather than creating an intermediate string.

Reply:
Yes, I added it.